File: 0735-asteroid-collision/0735-asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        int n = asteroids.length;
        int[] st = new int[n];
        int top = -1;

        for (int i = 0; i < n; i++) {
            if (asteroids[i] > 0) {
                st[++top] = asteroids[i];   // push
            } else {
                while (top != -1 && st[top] > 0 && st[top] < Math.abs(asteroids[i])) {
                    top--;   // pop
                }

                if (top != -1 && st[top] == Math.abs(asteroids[i])) {
                    top--;   // both destroy
                } else if (top == -1 || st[top] < 0) {
                    st[++top] = asteroids[i];   // push negative asteroid
                }
                // else: current asteroid gets destroyed, do nothing
            }
        }

        int[] result = new int[top + 1];
        for (int i = 0; i <= top; i++) {
            result[i] = st[i];
        }

        return result;
    }
}

File: 0735-asteroid-collision/0735-asteroid-collision.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
void push(int st[], int *top, int r){
    st[++(*top)]=r;
}
int pop(int st[],int *top){
    return st[(*top)--];
}
int* asteroidCollision(int* asteroids, int asteroidsSize, int* returnSize) {
    int n= asteroidsSize;
    int top=-1;
    int *st = (int *)malloc(n * sizeof(int));
    for(int i=0;i<n;i++){
        if(asteroids[i]>0){
            push(st,&top,asteroids[i]);
        }
        else{
            while(top!= -1 && st[top] > 0 && st[top] < abs(asteroids[i])){
                pop(st,&top);
            }
            if(top!=-1 && st[top]==abs(asteroids[i])){
                pop(st,&top);
            }
            else if(top==-1 || st[top]<0){
                push(st,&top,asteroids[i]);
            }
            else{
                continue;
            }
        }
    }
    *returnSize= top+1;
    return st;
}